update init stubs for Redis and redis Connection

cy opened the pull request:
This updates the `__init__` stubs for `redis.Redis` and `redis.Connection`, adding the missing params       `health_check_interval`, `client_name`, `username` to both, `single_connection_client` to `redis.Redis` and typing for those in `redis.Connection.__init__`.

Additionally, it looks like all the params in `redis.Redis.__init__` are optional, even those that aren't marked as such. (ie `host` defaults to `localhost`, `port` to 6379, `db` to 0). I've marked these as optional.

For reference https://github.com/andymccurdy/redis-py/blob/master/redis/connection.py#L496 https://github.com/andymccurdy/redis-py/blob/master/redis/connection.py#L496 
https://github.com/andymccurdy/redis-py/blob/master/redis/client.py#L737

working towards #3544

srittau commented:
Thanks. You added `Optional` to all `__init__` arguments. That doesn't look right to me. Please note that `Optional` is named a bit unfortunately. It is basically an alias for `Union[None, ...]`, not a marker for arguments with default values. For example, `socket_read_size` ultimately gets passed to `socket.recv()`, which doesn't accept `None` as value.

cy commented:
> Thanks. You added `Optional` to all `__init__` arguments. That doesn't look right to me. Please note that `Optional` is named a bit unfortunately. It is basically an alias for `Union[None, ...]`, not a marker for arguments with default values. For example, `socket_read_size` ultimately gets passed to `socket.recv()`, which doesn't accept `None` as value.

Thanks for correcting my misconception, @srittau. I've updated with `Optional` only for params whose default value is None. I also added `ssl_check_hostname`.

srittau commented on third_party/2and3/redis/client.pyi:
```diff
@@ -65,9 +65,14 @@ class Redis(object):
         ssl_keyfile: Optional[Text] = ...,
         ssl_certfile: Optional[Text] = ...,
-        ssl_cert_reqs: Optional[Union[str, int]] = ...,
+        ssl_cert_reqs: Union[str, int] = ...,
```
This was `Optional` before, and looking at the code of `SSLConnection`, there is an explicit `None` check, so the old version seems to be correct.

```suggestion
        ssl_cert_reqs: Optional[Union[str, int]] = ...,
```

cy replied:
Ah, I missed the None check. I've reverted to `Optional` here

srittau commented on third_party/2and3/redis/client.pyi:
```diff
@@ -65,9 +65,14 @@ class Redis(object):
+        ssl_check_hostname: bool = ...,
         max_connections: Optional[int] = ...,
+        single_connection_client: bool = ...,
+        health_check_interval: int = ...,
```
As this gets added to the result of `time()`, I expect that it also accepts `float` values:

```suggestion
        health_check_interval: float = ...,
```

cy replied:
updated!

srittau reviewed (changes requested):
Thanks, a few remarks below.
